### ISSUE TITLE
group together samples and exemplars in a single TimeSeries proto

### ISSUE THREAD

**Issue by cstyan** [prometheus/prometheus#8787]
Remote Write is currently doing the easiest but least efficient thing, which is allocating a new `prompb.TimeSeries` per sample or exemplar that is buffered, even if some samples/exemplars share the same series labels, in which case they could be put in the same `prompb.TimeSeries` for the write request. 

Nothing in the protocol or proto definitions currently stop us from making this improvement.

**Comment by bwplotka** [comment]
Also similar we could add for buffering samples together across series. As far as I can see: https://github.com/prometheus/prometheus/blob/8fd73b1d281d3227c337b291e8c797a85251d3d5/storage/remote/queue_manager.go#L525 we only add one sample to each write request.

**Comment by onprem** [comment]
> we only add one sample to each write request.

From what I understand, we are only adding the sample to a local queue. The requests only triggers when we either reach the configured max number of samples in a request or we exceed the batch send deadline.

https://github.com/prometheus/prometheus/blob/8fd73b1d281d3227c337b291e8c797a85251d3d5/storage/remote/queue_manager.go#L1055-L1127

**Comment by cstyan** [comment]
@bwplotka `Also similar we could add for buffering samples together across series.` There's limited scenarios in which this would ever actually come up. 

The first is if the user's Prometheus has less active series than `max_samples_per_send*max_shards`, which seems unlikely. 

The second is the replay/catch up scenario you brought up somewhere else. This one would be hard to get right. We'd have to balance buffering additional samples so we could slot them into the same `TimeSeries` with the right labels, with no using too much memory and OOM'ing the users Prometheus.

**Comment by Harkishen-Singh** [comment]
I think the reason we do not group exemplar and samples in a single Timeseries is that WAL for exemplar is different than the WAL for samples. If we store exemplar in the same sample's WAL, this will be easier to handle, given that both the WAL (exemplar and sample) have `ref` field.

Is there any reason why we didn't do that?

My guess is to keep WAL size reduced, but I don't think the increase will be significant.

**Comment by cstyan** [comment]
@Harkishen-Singh IIRC one of the main reasons was that modification of an existing WAL record type, such as Samples records, would have broken forwards/backwards compatibility between versions without some kind of converter tool.

Assume any improvements made as part of resolving this issue can't make changes to the WAL format.

**Comment by beorn7** [comment]
Hello from the bug scrub.

Nothing has moved here in the last 3y. We still think it's a valid idea, although we are not certain it will be a net win.
Maybe part of the overhead implied by this will already be done anyway in PRW2.0 as it has to create the symbol table.

Help still wanted. First step would be to do an MVP and benchmark it.

**Comment by cstyan** [comment]
Yep, in the current version of rw2.0 we still duplicate the label symbol IDs since we create a `TimeSeries` proto per sample/exemplar/etc. but the overall problem is reduced a lot by using the symbols over strings. Still worth trying something here, but not an immediate priority.
